Race condition in restarting folder

`TestIntroducer` triggers this:

```
10:29:43 INFO: Connection to AIR6LPZ-7K4PTTV-UXQSMUU-CPQ5YWH-OEDFIIQ-JUG777G-2YQXXR5-YD6AWQR at  closed: restarting folder folder1
10:29:43 INFO: Connection to GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY at  closed: restarting folder folder1
10:29:43 INFO: Restarted folder folder1 (sendreceive)
10:29:43 INFO: No stored folder metadata for "folder2": recalculating
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb60084]

goroutine 1150 [running]:
github.com/syncthing/syncthing/lib/db.(*FileSet).ListDevices(0x0, 0x1, 0x1, 0xc0005e63c0)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/db/set.go:367 +0x34
github.com/syncthing/syncthing/lib/model.(*model).startFolderLocked(0xc005466b00, 0xe4c442, 0x7, 0x0, 0x0, 0x0, 0xe4cd3b, 0x8, 0x0, 0xc00557b3c0, ...)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/model/model.go:271 +0x226
github.com/syncthing/syncthing/lib/model.(*model).RestartFolder(0xc005466b00, 0xe4c442, 0x7, 0x0, 0x0, 0x0, 0xe4cd3b, 0x8, 0x0, 0xc0003f4a80, ...)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/model/model.go:484 +0x930
github.com/syncthing/syncthing/lib/model.(*model).CommitConfiguration(0xc005466b00, 0x0, 0xc005b39b80, 0x2, 0x2, 0xc005b3af00, 0x2, 0x2, 0x0, 0x0, ...)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/model/model.go:2534 +0xd26
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListener(0xc004f60800, 0xf92aa0, 0xc005466b00, 0x0, 0xc005b39b80, 0x2, 0x2, 0xc005b3af00, 0x2, 0x2, ...)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/config/wrapper.go:241 +0x1d0
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners.func1(0xc004f60800, 0xc000345180, 0xc000345500, 0xf93fa0, 0xc000226ec0, 0xf92aa0, 0xc005466b00)
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/config/wrapper.go:232 +0x12b
created by github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners
	/opt/tcagent/syncthing-4-work/src/github.com/syncthing/syncthing/lib/config/wrapper.go:231 +0x28e
FAIL	github.com/syncthing/syncthing/lib/model	3.693s
```

Theoretically there should be a file set but apparently it's nil (deleted). There's no locking to prevent this as we release the lock in RestartFolder between creating the set and calling startFolderLocked.

I completely forgot about this. Now that https://github.com/syncthing/syncthing/pull/6107 is merged I get the following panics when running tests on master, which can be explained by the same thing:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xe6581a]

goroutine 821 [running]:
github.com/syncthing/syncthing/lib/db.(*FileSet).ListDevices(...)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/db/set.go:367
github.com/syncthing/syncthing/lib/model.(*model).startFolderLocked(0xc0002fde40, 0x10e74aa, 0x7, 0x0, 0x0, 0x0, 0x10e7e60, 0x8, 0x0, 0xc0004ff500, ...)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:301 +0x20a
github.com/syncthing/syncthing/lib/model.(*model).restartFolder(0xc0002fde40, 0x10e74aa, 0x7, 0x0, 0x0, 0x0, 0x10e7e60, 0x8, 0x0, 0xc0004ff480, ...)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:524 +0xa25
github.com/syncthing/syncthing/lib/model.(*model).CommitConfiguration(0xc0002fde40, 0x0, 0xc00068a840, 0x2, 0x2, 0xc00000de00, 0x2, 0x2, 0x0, 0x0, ...)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:2610 +0xdb5
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListener(0xc00017ac00, 0x127d080, 0xc0002fde40, 0x0, 0xc00068a840, 0x2, 0x2, 0xc00000de00, 0x2, 0x2, ...)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/config/wrapper.go:211 +0x1d0
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners.func1(0xc00017ac00, 0xc000035500, 0xc000035880, 0x127db80, 0xc00003a9d0, 0x127d080, 0xc0002fde40)
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/config/wrapper.go:202 +0x12b
created by github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners
	/media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/config/wrapper.go:201 +0x286
FAIL	github.com/syncthing/syncthing/lib/model	6.996s
```

Do you already have another error-bubble-up-PR incoming soon that can address this? Otherwise we need to put some kind of band-aid on this for the next release.

--- a/lib/config/wrapper.go
+++ b/lib/config/wrapper.go
@@ -94,6 +94,7 @@ type wrapper struct {
 	path     string
 	evLogger events.Logger
 
+	waiter    Waiter // Latest ongoing config change
 	deviceMap map[protocol.DeviceID]DeviceConfiguration
 	folderMap map[string]FolderConfiguration
 	subs      []Committer
@@ -109,6 +110,7 @@ func Wrap(path string, cfg Configuration, evLogger events.Logger) Wrapper {
 		cfg:      cfg,
 		path:     path,
 		evLogger: evLogger,
+		waiter:   noopWaiter{}, // Noop until first config change
 		mut:      sync.NewMutex(),
 	}
 	return w
@@ -144,7 +146,8 @@ func (w *wrapper) Subscribe(c Committer) {
 }
 
 // Unsubscribe de-registers the given handler from any future calls to
-// configuration changes
+// configuration changes and only returns after a potential ongoing config
+// change is done.
 func (w *wrapper) Unsubscribe(c Committer) {
 	w.mut.Lock()
 	for i := range w.subs {
@@ -155,7 +158,11 @@ func (w *wrapper) Unsubscribe(c Committer) {
 			break
 		}
 	}
+	waiter := w.waiter
 	w.mut.Unlock()
+	// Waiting mustn't be done under lock, as the goroutines in notifyListener
+	// may dead-lock when trying to access lock on config read operations.
+	waiter.Wait()
 }
 
 // RawCopy returns a copy of the currently wrapped Configuration object.
@@ -191,7 +198,9 @@ func (w *wrapper) replaceLocked(to Configuration) (Waiter, error) {
 	w.deviceMap = nil
 	w.folderMap = nil
 
-	return w.notifyListeners(from.Copy(), to.Copy()), nil
+	w.waiter = w.notifyListeners(from.Copy(), to.Copy())
+
+	return w.waiter, nil
 }
 
 func (w *wrapper) notifyListeners(from, to Configuration) Waiter {

--- a/lib/connections/service.go
+++ b/lib/connections/service.go
@@ -193,6 +193,12 @@ func NewService(cfg config.Wrapper, myID protocol.DeviceID, mdl Model, tlsCfg *t
 	return service
 }
 
+func (s *service) Stop() {
+	s.cfg.Unsubscribe(s.limiter)
+	s.cfg.Unsubscribe(s)
+	s.Supervisor.Stop()
+}
+
 func (s *service) handle(ctx context.Context) {
 	var c internalConn
 	for {

--- a/lib/model/model.go
+++ b/lib/model/model.go
@@ -217,7 +217,6 @@ func NewModel(cfg config.Wrapper, id protocol.DeviceID, clientName, clientVersio
 	}
 	m.Add(m.progressEmitter)
 	scanLimiter.setCapacity(cfg.Options().MaxConcurrentScans)
-	cfg.Subscribe(m)
 
 	return m
 }
@@ -241,9 +240,11 @@ func (m *model) onServe() {
 		}
 		m.newFolder(folderCfg)
 	}
+	m.cfg.Subscribe(m)
 }
 
 func (m *model) Stop() {
+	m.cfg.Unsubscribe(m)
 	m.Supervisor.Stop()
 	devs := m.cfg.Devices()
 	ids := make([]protocol.DeviceID, 0, len(devs))

--- a/lib/model/progressemitter.go
+++ b/lib/model/progressemitter.go
@@ -23,6 +23,7 @@ import (
 type ProgressEmitter struct {
 	suture.Service
 
+	cfg                config.Wrapper
 	registry           map[string]map[string]*sharedPullerState // folder: name: puller
 	interval           time.Duration
 	minBlocks          int
@@ -40,6 +41,7 @@ type ProgressEmitter struct {
 // DownloadProgress events every interval.
 func NewProgressEmitter(cfg config.Wrapper, evLogger events.Logger) *ProgressEmitter {
 	t := &ProgressEmitter{
+		cfg:                cfg,
 		registry:           make(map[string]map[string]*sharedPullerState),
 		timer:              time.NewTimer(time.Millisecond),
 		sentDownloadStates: make(map[protocol.DeviceID]*sentDownloadState),
@@ -51,7 +53,6 @@ func NewProgressEmitter(cfg config.Wrapper, evLogger events.Logger) *ProgressEmi
 	t.Service = util.AsService(t.serve, t.String())
 
 	t.CommitConfiguration(config.Configuration{}, cfg.RawCopy())
-	cfg.Subscribe(t)
 
 	return t
 }
@@ -59,6 +60,9 @@ func NewProgressEmitter(cfg config.Wrapper, evLogger events.Logger) *ProgressEmi
 // serve starts the progress emitter which starts emitting DownloadProgress
 // events as the progress happens.
 func (t *ProgressEmitter) serve(ctx context.Context) {
+	t.cfg.Subscribe(t)
+	defer t.cfg.Unsubscribe(t)
+
 	var lastUpdate time.Time
 	var lastCount, newCount int
 	for {

--- a/lib/ur/usage_report.go
+++ b/lib/ur/usage_report.go
@@ -58,7 +58,6 @@ func New(cfg config.Wrapper, m model.Model, connectionsService connections.Servi
 		forceRun:           make(chan struct{}, 1), // Buffered to prevent locking
 	}
 	svc.Service = util.AsService(svc.serve, svc.String())
-	cfg.Subscribe(svc)
 	return svc
 }
 
@@ -385,6 +384,9 @@ func (s *Service) sendUsageReport() error {
 }
 
 func (s *Service) serve(ctx context.Context) {
+	s.cfg.Subscribe(s)
+	defer s.cfg.Unsubscribe(s)
+
 	t := time.NewTimer(time.Duration(s.cfg.Options().URInitialDelayS) * time.Second)
 	for {
 		select {

